AI #2359: Add Actor (`Consumer ID` and `Principal ID`) columns to `Cost and Usage` dataset (#2360)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/appendix/actor_attribution.md

# Examples: Actor Attribution

> The examples below illustrate how [Principal ID](#datasets.costandusage.principalid) and [Consumer ID](#datasets.costandusage.consumerid) in the [Cost and Usage](#datasets.costandusage) dataset are populated across different data generators and technology categories to resolve asymmetric actor granularity. Comparing these scenarios demonstrates how FOCUS handles attribution when the infrastructure actor and application actor are distinct, identical, or mutually exclusive.

Reviewer: ```suggestion
The examples below illustrate how [Principal ID](#datasets.costandusage.principalid) and [Consumer ID](#datasets.costandusage.consumerid) in the [Cost and Usage](#datasets.costandusage) dataset are populated across different scenarios, data generators, and technology categories to resolve asymmetric actor granularity. Comparing these scenarios demonstrates how FOCUS represents the [*principal*](#glossary:principal) that initiated a request and the [*consumer*](#glossary:consumer) that consumed the [*resource*](#glossary:resource) or [*service*](#glossary:service) when the *principal* and *consumer* are distinct, identical, or either is not populated.
```

Author: A tweak of your suggestion.

```suggestion
The examples below illustrate how [Principal ID](#datasets.costandusage.principalid) and [Consumer ID](#datasets.costandusage.consumerid) in the [Cost and Usage](#datasets.costandusage) dataset resolve asymmetric actor granularity. These scenarios demonstrate how FOCUS represents the [*principal*](#glossary:principal) and [*consumer*](#glossary:consumer) of a [*resource*](#glossary:resource) or [*service*](#glossary:service) when these actors are distinct, identical, or omitted.
```

1. **Generative AI API:** **Acme Corp** uses an internal bot to summarize notes via a LatticeScale API. Security audits the service account (`PrincipalId`), while FinOps allocates token costs to the employee (`ConsumerId`).
2. **Multi-Tenant PaaS:** A shared BI engine runs a query on OmniQuery for a specific client of **GearPeak Outdoors**. Capturing the downstream client (`ConsumerId`) prevents costs from pooling in overhead.

> 3. **Direct IaaS:** A data engineer at **AeroScale** directly provisions an Aura Web compute cluster. Because the infrastructure is consumed directly by the principal, no downstream application actor exists.

Reviewer: My baseline assumption in reviewing the Principal ID and Consumer ID model is that the two columns are not differentiated by infrastructure vs application layers, but rather by their role in the transaction: PrincipalId represents the actor initiating the request, while ConsumerId represents the downstream actor consuming the resource or service. Based on this assumption, I have proposed corresponding updates to the glossary definitions, introductory sections, and nullability requirements.

If this interpretation is adopted, I suggest revising Scenario 3:

```suggestion
3. **Network Edge Processing:** At **Acme Corp**, traffic routed through an Aura Web edge network is authenticated at the gateway (`PrincipalId`), but downstream consuming entities cannot be identified due to aggregated network traffic.
```

Author: @ijurica 
Instead of piecemeal comments, would you mind issuing a PR against this branch that aggregates all your proposed appendix changes?

Author: @ijurica has proposed PR #2401.  We will consider that one, merge as desired, and then resolve this comment.

Author: Resolving now that we have merged #2401.

> 4. **Seat-Based SaaS:** **Acme Corp** pays for SprintCanvas project management licenses. The provider does not attribute an infrastructure actor, but the opaque `ConsumerId` allows safe user-level chargeback.

Reviewer: ```suggestion
4. **Seat-Based SaaS:** **Acme Corp** pays for SprintCanvas project management licenses. The provider cannot associate charges with a [*principal*](#glossary:principal), but the opaque `ConsumerId` allows safe user-level chargeback.
```

Author: Going to hold on making any further changes to conditionality of nullability until there are wider TF and Maintainer discussions about this.

Author: @ijurica has proposed PR #2401.  We will consider that one, merge as desired, and then resolve this comment.

Author: Resolving now that we have merged #2401.

> 5. **Direct PaaS Usage:** A data scientist at **Acme Corp** logs into a LatticeScale managed notebook using their individual SSO credential. Because the user authenticated directly to the platform and is the sole consumer of the compute, the same opaque identifier is populated in both columns.

Reviewer 2: Proposing a 6th scenario to demonstrate that the dual-column model handles vendor pricing-model transitions over time, not just static patterns. The seat-to-token shift documented in [SemiAnalysis](https://semianalysis.com/tokenomics-model/) (Salesforce and Adobe cited by name as adding per-token billing on top of seat-based pricing for embedded AI features) is accommodated by the existing `PrincipalId` / `ConsumerId` model without schema changes, and a 6th scenario makes that explicit. See [FR #2358](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/2358) for the full use case write-up.

Recognizing this expands the PR's original 5-scenario scope by one; happy to split into a follow-up PR if you'd rather keep this one tight.

```suggestion
5. **Direct PaaS Usage:** A data scientist at **Acme Corp** logs into a LatticeScale managed notebook using their individual SSO credential. Because the user authenticated directly to the platform and is the sole consumer of the compute, the same opaque identifier is populated in both columns.
6. **Hybrid Seat + Token SaaS:** **Acme Corp** subscribes to PipelCRM and uses its embedded AI assistant. The seat license and per-token AI usage post as separate line items under the same opaque `ConsumerId`, so chargeback follows the user as vendors shift from seat-only to seat + consumption pricing.
```

Author: @Matt-Cowsert Yes -- could you please issue a separate PR against this branch?  I think it'll make it far clearer what you are proposing, and then I can integrate.

Author: @Matt-Cowsert has proposed PR #2400.  We will consider that one, merge as desired, and then resolve this comment.

Author: Resolving this thread in favor of PR #2400.

> | Scenario | Technology Category | Data Generator | PrincipalId | ConsumerId |

Reviewer: Consider renaming the second column to “Scenario” to avoid conflict with FinOps Technology Categories.

```suggestion
| # | Scenario | Data Generator | PrincipalId | ConsumerId |
```

| :--- | :--- | :--- | :--- | :--- |
| 1 | **Generative AI API** | LatticeScale | `svc-acme-docbot-prod` | `emp_hash_84729x` |
| 2 | **Multi-Tenant PaaS** | OmniQuery | `svc-bi-reporting-engine` | `client_gearpeak_001` |

> | 3 | **Direct IaaS** | Aura Web | `user_d_engineer_993` | `null` |

Reviewer: ```suggestion
| 3 | **Network Edge Processing** | Aura Web | `svc-edge-gateway-prod` | `null` |
```

Author: Not sure about this change -- how come?

Author: @ijurica has proposed PR #2401.  We will consider that one, merge as desired, and then resolve this comment.

Author: Resolving now that we have merged #2401.

| 4 | **Seat-Based SaaS** | SprintCanvas | `null` | `user_uuid_554321` |

> | 5 | **Direct PaaS Usage** | LatticeScale | `dev_uuid_112233` | `dev_uuid_112233` |

Reviewer 2: Companion table row for the new "Hybrid Seat + Token SaaS" scenario, which demonstrates the dual-column model handles vendor pricing-model transitions (e.g., the seat-to-token shift documented in [SemiAnalysis](https://semianalysis.com/tokenomics-model/)). Same opaque `ConsumerId` carries chargeback across both the seat-license line and the per-token AI overage line for the same user. See [FR #2358](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/2358) for the full use case write-up.

```suggestion
| 5 | **Direct PaaS Usage** | LatticeScale | `dev_uuid_112233` | `dev_uuid_112233` |
| 6 | **Hybrid Seat + Token SaaS** | PipelCRM | `null` | `user_uuid_887766` |
```

Author: @Matt-Cowsert Yes -- could you please issue a separate PR against this branch?  I think it'll make it far clearer what you are proposing, and then I can integrate.

Author: @Matt-Cowsert has proposed PR https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2400. We will consider that one, merge as desired, and then resolve this comment.

Author: Resolving this thread in favor of PR #2400.

# Examples: Actor Attribution

The examples below illustrate how [Principal ID](#datasets.costandusage.principalid) and [Consumer ID](#datasets.costandusage.consumerid) in the [Cost and Usage](#datasets.costandusage) dataset resolve asymmetric actor granularity. These scenarios demonstrate how FOCUS represents the [*principal*](#glossary:principal) and [*consumer*](#glossary:consumer) of a [*resource*](#glossary:resource) or [*service*](#glossary:service) when these actors are distinct, identical, or omitted.

> 1. **Generative AI API:** **Acme Corp** uses an internal bot to summarize notes via a LatticeScale API. Security audits the service account (`PrincipalId`), while FinOps allocates token costs to the employee (`ConsumerId`).

Reviewer 3: I've read through the PR and I'm still trying to understand where it makes sense to have both of these. I think this line illustrates the challenge I have: In what situation would a provider be able to give both Principal and Consumer info?

In this scenario, I'm imagining using an LLM Gateway (principal = service account) which users then call (consumer = user). However, the provider will never see the user/consumer because it's abstracted by the LLM Gateway.

Lets say that another company (or division of the same company) lets their users consume the LLM directly instead of through a gateway. How would the provider be able to differentiate?

Reviewer 2: 🤖 [AI][Claude Code] The Actor Attribution appendix has moved out of this PR. Per the TF2 and Maintainers decision to deliver the actor columns incrementally, #2360 is now scoped to `PrincipalId` only, and the both-column appendix (including this scenario) has been carved out to draft PR #2495, stacked on this branch.

This question about when a provider can surface both `PrincipalId` and `ConsumerId`, including the LLM-gateway case where the end consumer is abstracted away, belongs with the `ConsumerId` definition, so let us carry it over to #2495: https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2495

> 1. **Generative AI API:** Acme Corp uses an internal bot to summarize notes via a LatticeScale API. The bot authenticates as a service account, which LatticeScale records as the *principal* (`PrincipalId`). The employee whose notes are summarized benefits from the *charge* and is not the *principal*.

Reviewer 4: Do we need to call out who benefits in each of these scenarios, given that we've said above that the beneficiary isn't really related to who we're capturing as a principal/credential?

Reviewer 2: Addressed in #2636. The per-scenario beneficiary notes are gone, and the Attribution Roles section states once that neither column identifies the party that benefits from a charge.

5. **Seat-Plus-Token SaaS:** Acme Corp subscribes to PipelCRM and uses its embedded AI assistant. The seat license and the per-token AI usage are reported as separate *charges* under the same seat holder (`PrincipalId`), who also benefits from both *charges*, so attribution follows the user as *service providers* shift from seat-based to hybrid seat and consumption-based pricing.
6. **Direct PaaS Usage:** A data scientist at Acme Corp logs into a LatticeScale managed notebook using their individual SSO credential. LatticeScale grants access to that user identity, which is the *principal* (`PrincipalId`), and the same user benefits from the *charge*.
7. **Billing System Charge:** Acme Corp receives a promotional credit or is charged applicable tax from LatticeScale. The *charge* originates in the *service provider's* billing system and is not associated with any entity in its identity and access management model, so `PrincipalId` is `null`.

> | # | Scenario | Data Generator | PrincipalId |

Reviewer 4: This table really has no meaning/value IMO. No one can correlate these values to anything they would understand.

Reviewer 2: Your rewrite in #2636 replaced this table. It now carries both identifiers and a Credential Type column, with the Attribution Roles section above it stating what each column answers.
